pkg/miniogw/multipart.go

```diff
@@ -0,0 +1,438 @@
+	defer mon.Task()(&ctx)(&err)
+
+	uploads := s.storj.multipart
+
+	upload, err := uploads.Get(bucket, object, uploadID)
+	if err != nil {
+		return minio.PartInfo{}, err
+	}
+
+	part, err := upload.Stream.AddPart(partID, data)
+	if err != nil {
+		return minio.PartInfo{}, err
+	}
+
+	err = <-part.Done
+	if err != nil {
+		return minio.PartInfo{}, err
+	}
+
+	partInfo := minio.PartInfo{
+		PartNumber:   part.Number,
+		LastModified: time.Now(),
+		ETag:         data.SHA256HexString(),
+		Size:         atomic.LoadInt64(&part.Size),
+	}
+
+	upload.mu.Lock()
+	upload.completed = append(upload.completed, partInfo)
+	upload.mu.Unlock()
+
+	return partInfo, nil
+}
+
+func (s *storjObjects) AbortMultipartUpload(ctx context.Context, bucket, object, uploadID string) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	uploads := s.storj.multipart
+
+	upload, err := uploads.Remove(bucket, object, uploadID)
+	if err != nil {
+		return err
+	}
+
+	errAbort := Error.New("abort")
+	upload.Stream.Abort(errAbort)
+	r := <-upload.Done
+	if r.Error != errAbort {
+		return r.Error
+	}
+	return nil
+}
+
+func (s *storjObjects) CompleteMultipartUpload(ctx context.Context, bucket, object, uploadID string, uploadedParts []minio.CompletePart) (objInfo minio.ObjectInfo, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	uploads := s.storj.multipart
+	upload, err := uploads.Remove(bucket, object, uploadID)
+	if err != nil {
+		return minio.ObjectInfo{}, err
+	}
+
+	// notify stream that there aren't more parts coming
+	upload.Stream.Close()
+	// wait for completion
+	result := <-upload.Done
+	// return the final info
+	return result.Info, result.Error
+}
+
+// func (s *storjObjects) ListMultipartUploads(ctx context.Context, bucket, prefix, keyMarker, uploadIDMarker, delimiter string, maxUploads int) (result minio.ListMultipartsInfo, err error) {
+// 	return
+// }
+
+func (s *storjObjects) ListObjectParts(ctx context.Context, bucket, object, uploadID string, partNumberMarker int, maxParts int) (result minio.ListPartsInfo, err error) {
+	uploads := s.storj.multipart
+	upload, err := uploads.Get(bucket, object, uploadID)
+	if err != nil {
+		return minio.ListPartsInfo{}, err
+	}
+
+	list := minio.ListPartsInfo{}
+
+	list.Bucket = bucket
+	list.Object = object
+	list.UploadID = uploadID
+	list.PartNumberMarker = partNumberMarker
+	list.MaxParts = maxParts
+	list.UserDefined = upload.Metadata
+
+	upload.mu.Lock()
+	list.Parts = append(list.Parts, upload.completed...)
+	upload.mu.Unlock()
+
+	sort.Slice(list.Parts, func(i, k int) bool {
+		return list.Parts[i].PartNumber < list.Parts[k].PartNumber
+	})
+
+	var first int
+	for i, p := range list.Parts {
+		first = i
+		if partNumberMarker <= p.PartNumber {
+			break
+		}
+	}
+
+	list.Parts = list.Parts[first:]
+	if len(list.Parts) > maxParts {
+		list.NextPartNumberMarker = list.Parts[maxParts+1].PartNumber
```

Comment:
I think that `+1` is wrong. On the next line we are taking `[:maxParts]`, which means from 0 to `maxParts-1` inclusively. So the next marker should be set to `maxParts`, not `maxPart+1`, shouldn't it?

Reply:
It needs to return the next part number, so the +1 part is correct. However it doesn't handle the edge case properly. Good catch.

Reply:
Ah, yes... it should be +1. You were right.